File: src/main/java/io/github/emcw/map/api/Residents.java
package io.github.emcw.map.api;

import com.github.benmanes.caffeine.cache.Cache;

import io.github.emcw.caching.BaseCache;
import io.github.emcw.caching.CacheOptions;
import io.github.emcw.map.entities.Resident;
import io.github.emcw.exceptions.MissingEntryException;
import io.github.emcw.utils.parsers.SquaremapParser;

import org.jetbrains.annotations.NotNull;

import java.util.Map;

public class Residents extends BaseCache<Resident> {
    public Residents(CacheOptions options) {
        super(options);

        setUpdater(this::forceUpdate);
        forceUpdate();

        build();
    }

    public void tryUpdate() {
        tryExpire();
        updateCache(false);
    }

    public void forceUpdate() {
        updateCache(true);
    }

    private void updateCache(Boolean force) {
        if (!empty() && !force) return;

        // Parse player data into usable Player objects.
        SquaremapParser.parseMapData(false, false, true);
        Cache<String, Resident> residents = SquaremapParser.getParsedResidents();

        // Make sure we're using valid data to populate the cache with.
        if (residents == null) return;
        if (residents.asMap().isEmpty()) return;

        setCache(residents);
    }

    @Override
    public Map<String, Resident> all() {
        tryUpdate();
        return super.all();
    }

    @Override
    public Resident single(String name) throws MissingEntryException {
        tryUpdate();
        return super.single(name);
    }

    public Map<String, Resident> get(String @NotNull ... keys) {
        tryUpdate();
        return super.get(keys);
    }
}

File: src/main/kotlin/io/github/emcw/map/api/Residents.kt
package io.github.emcw.map.api

import io.github.emcw.EMCMap
import io.github.emcw.caching.BaseCache
import io.github.emcw.caching.CacheOptions
import io.github.emcw.exceptions.MissingEntryException
import io.github.emcw.map.entities.Resident
import io.github.emcw.utils.DataParser.parseMapData
import io.github.emcw.utils.DataParser.parsedResidents

class Residents(private val parent: EMCMap, options: CacheOptions) : BaseCache<Resident?>(options) {
    init {
        setUpdater { forceUpdate() }
        forceUpdate()
        build()
    }

    fun tryUpdate() {
        tryExpire()
        updateCache(false)
    }

    fun forceUpdate() {
        updateCache(true)
    }

    private fun updateCache(force: Boolean) {
        if (!empty() && !force) return

        // Parse player data into usable Player objects.
        parseMapData(parent.mapName, false, false, true)
        val residents = parsedResidents()

        // Make sure we"re using valid data
        if (!residents.asMap().isEmpty()) setCache(residents)
    }

    override fun all(): Map<String?, Resident?>? {
        tryUpdate()
        return super.all()
    }

    @Throws(MissingEntryException::class)
    override fun single(name: String?): Resident? {
        tryUpdate()
        return super.single(name)
    }

    override fun get(vararg keys: String): Map<String?, Resident?>? {
        tryUpdate()
        return super.get(*keys)
    }
}